### PR TITLE
feat: using enum_dispatch to represent data_type(DataTypePtr -> DataTypeImpl)

### LINKED ISSUE #4868: Feature: using enum_dispatch to represent data_type
**Summary**

Description for this feature.

https://docs.rs/enum_dispatch/latest/enum_dispatch/index.html

Pros to use `enum_dispatch`:

1. Better performance (static dispatch, [compiler_optimized](https://docs.rs/enum_dispatch/latest/enum_dispatch/index.html#compiler_optimized)).
2. Avoid using `typetag::serde` to serde.
3. Automatic `From` and `TryInto` implementations (No downcast  unwrap anymore)
4. attribute support for enum

Tasks:
- [ ] DataTypePtr -> DataTypeImpl (Maybe a better name)
- [ ] ColumnRef -> ColumnImpl
- [ ] `Box<TypeSerializer>` -> TypeSerializerImpl
- [ ] `Box<TypeDeserializer>`  -> TypeDeserializerImpl

These tasks could lead many codes to be refactored.  You can pick it if you enjoy refactoring codes :)

Performance in databend:

```
data_type_ptr_create              time:   [52.972 ns 53.001 ns 53.033 ns]                                 
data_type_enum_create              time:   [53.507 ns 53.535 ns 53.563 ns]                                 

data_type_ptr_dummy               time:   [3.3817 ns 3.3830 ns 3.3843 ns]                                 
data_type_enum_dummy              time:   [430.82 ps 430.94 ps 431.09 ps]                                 

data_type_ptr_dummy2              time:   [1.0795 ns 1.0803 ns 1.0811 ns]                                  
data_type_enum_dummy2             time:   [430.57 ps 430.61 ps 430.66 ps]                                  

data_type_ptr_dummy2_compiled     time:   [1.2825 ns 1.2827 ns 1.2829 ns]
data_type_enum_dummy2_compiled    time:   [214.03 ps 214.06 ps 214.10 ps]
```

### PR DESCRIPTION
**Pull request by PsiACE**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

This is a breaking change. 

- First, we introduced `DataTypeImpl` to replace the original `DataTypePtr`. 
- Second, the serialized name of our type lost the `Type` suffix. 
- Most of the time, the name we passed also used `data_type.name( )` .

## Changelog

- Improvement
- Performance Improvement

## Related Issues

Part of #4868

### CODE REVIEW

File: common/meta/types/tests/it/compatible.rs [hunk 1]
```diff
@@ -20,6 +20,7 @@ use openraft::raft::Entry;
 use openraft::raft::EntryPayload;
 
 #[test]
+#[ignore]
 fn test_load_entry_compatibility() -> anyhow::Result<()> {
```
[the remark this answers is not in the document]
Comment (drmingdrmer) [reply common/meta/types/tests/it/compatible.rs:24]: It's for backward compatibility and according to a recent discussion, this test can be just ignored.

The data type compatibility will be done at category level, metasrv will never need to do this job.


@BohuTANG right?

[the remark this answers is not in the document]
Comment (drmingdrmer) [reply common/meta/types/tests/it/compatible.rs:24]: > Also, please help check the metactl tests.

1. remove all `_meta*`.
2. run stateless test to collect some data in metasrv.
3. export all meta data from one of the `_meta*` dir with `./target/debug/databend-metactl --export --raft-dir "$meta_dir" > $exported`.
4. replace `tests/metactl/meta.json` withe exported file.

[the remark this answers is not in the document]
Comment (PsiACE) [reply common/meta/types/tests/it/compatible.rs:24]: > It's for backward compatibility and according to a recent discussion, this test can be just ignored.

So, should i delete it or keep `#[ignore]`

[the remark this answers is not in the document]
Comment (drmingdrmer) [reply common/meta/types/tests/it/compatible.rs:24]: Ignore it if other tests all pass

File: common/datavalues/src/types/type_primitive.rs [hunk 2]
```diff
@@ -32,20 +32,20 @@ pub struct PrimitiveDataType<
...
 
 // typetag did not support generic impls, so we have to do this
-pub fn create_primitive_datatype<T: PrimitiveType>() -> Arc<dyn DataType> {
+pub fn create_primitive_datatype<T: PrimitiveType>() -> DataTypeImpl {
     match (T::SIGN, T::FLOATING, T::SIZE) {
-        (false, false, 1) => Arc::new(UInt8Type { _t: PhantomData }),
-        (false, false, 2) => Arc::new(UInt16Type { _t: PhantomData }),
-        (false, false, 4) => Arc::new(UInt32Type { _t: PhantomData }),
-        (false, false, 8) => Arc::new(UInt64Type { _t: PhantomData }),
+        (false, false, 1) => DataTypeImpl::UInt8(UInt8Type { _t: PhantomData }),
```
Comment (Xuanwo) [inline common/datavalues/src/types/type_primitive.rs:37]: Can we use `UInt8Type::default()` here?

File: common/datavalues/src/types/type_primitive.rs [hunk 3]
```diff
@@ -54,8 +54,8 @@ pub fn create_primitive_datatype<T: PrimitiveType>() -> Arc<dyn DataType> {
 macro_rules! impl_numeric {
     ($ty:ident, $tname:ident, $name: expr, $sql_name:expr, $alias: expr) => {
         impl PrimitiveDataType<$ty> {
-            pub fn arc() -> DataTypePtr {
-                Arc::new(Self { _t: PhantomData })
+            pub fn arc() -> DataTypeImpl {
+                DataTypeImpl::$tname(Self { _t: PhantomData })
```
Comment (Xuanwo) [inline common/datavalues/src/types/type_primitive.rs:58]: Same question

File: common/datavalues/src/types/type_variant_object.rs [hunk 4]
```diff
@@ -27,12 +27,11 @@ use crate::prelude::*;
 pub struct VariantObjectType {}
```
Comment (Xuanwo) [inline common/datavalues/src/types/type_variant_object.rs:27]: Same question.

File: common/datavalues/src/types/type_primitive.rs (hunk 2 above)
Comment (sundy-li) [reply common/datavalues/src/types/type_primitive.rs:37]: Seems it could, I did not notice that `PhantomData ` default impls.

Comment (PsiACE) [reply common/datavalues/src/types/type_primitive.rs:37]: We can continue to improve these in other PRs if feasible.

**Review by Xuanwo — APPROVED** [review]

**Review by sundy-li — APPROVED** [review]

File: common/datavalues/src/types/type_coercion.rs [hunk 5]
```diff
@@ -323,7 +322,7 @@ pub fn merge_types(lhs_type: &DataTypePtr, rhs_type: &DataTypePtr) -> Result<Dat
             let b = rhs_type.as_any().downcast_ref::<ArrayType>().unwrap();
```
Comment (sundy-li) [inline common/datavalues/src/types/type_coercion.rs:322]: I suggest to remove all `downcast_ref`, since we have `try_from`, `try_into` now!

**Comment by Veeupup** [comment]
@PsiACE I'm interested that how much performence we get by using enum_dispatch : D

**Comment by sundy-li** [comment]
>  I'm interested that how much performence we get by using enum_dispatch : D

It's not the hot path, just to make the codes better.

**Pull request merged**